ssl/ssl_sess.c

```diff
@@ -1145,11 +1237,40 @@ static void SSL_SESSION_list_add(SSL_CTX *ctx, SSL_SESSION *s)
         s->prev = (SSL_SESSION *)&(ctx->session_cache_head);
         s->next = (SSL_SESSION *)&(ctx->session_cache_tail);
     } else {
-        s->next = ctx->session_cache_head;
-        s->next->prev = s;
-        s->prev = (SSL_SESSION *)&(ctx->session_cache_head);
-        ctx->session_cache_head = s;
+        if (timeoutcmp(s, ctx->session_cache_head) >= 0) {
+            /*
+             * if we timeout after (or the same time as) the first
+             * session, put us first - usual case
+             */
+            s->next = ctx->session_cache_head;
+            s->next->prev = s;
+            s->prev = (SSL_SESSION *)&(ctx->session_cache_head);
+            ctx->session_cache_head = s;
+        } else if (timeoutcmp(s, ctx->session_cache_tail) < 0) {
+            /* if we timeout before the last session, put us last */
+            s->prev = ctx->session_cache_tail;
+            s->prev->next = s;
+            s->next = (SSL_SESSION *)&(ctx->session_cache_tail);
+            ctx->session_cache_tail = s;
+        } else {
+            /*
+             * we timeout somewhere in-between - if there is only
+             * one session in the cache it will be caught above
+             */
+            next = ctx->session_cache_head->next;
+            while (next != (SSL_SESSION*)&(ctx->session_cache_tail)) {
```

Comment:
Quadratic time here.  Do we care?

A heap (priority queue) would be the appropriate algorithm.

Reply:
How are you getting O(N^2)? This is O(N), given the single loop.

That would be a significant rewrite for the SSL_SESSION object.

Reply:
O(n) for each of n items put into the list which means O(n²) in total.
A heap would be O(log n) for each item meaning O(n log n) total.

Reply:
There is no operation that adds (n) sessions at one time... ;)
But, for new sessions, they will hit one of the first conditions above, which adds directly to the start or end of the list, which is just O(1); no worse than the existing code.
Existing sessions that have their time updated to "now", will also be added at O(1).
Existing sessions that _do not_ have their time updated will be added at O(n).

So, when the optimization is used, it's still O(n) when adding (n) items.

When sessions need to be timed out (either every 255 sessions, or when the number of sessions exceeds the threshold - I'm not looking at that code right now) or flushed manually), that operation is reduced from O(n) to O(fractional n), since the search ends early.

Again, a significant rewrite of the SSL_SESSION object for a heap (is there a HEAP storage structure in OpenSSL?).